### PR TITLE
Add support for cross-site authentication.

### PR DESCRIPTION
**Pull request by jamestmartin**
There was previously a different mechanism here (describing what's implemented so far), but I have a better mechanism so it's obsolete. If you're curious what it was, check the edit history.

### CODE REVIEW

**Comment by jamestmartin** [comment]
Oh yeah, if verify invalidates the auth code, it should also be a POST rather than a GET because it is potent. That sounds like another problem for tomorrow.

**Comment by jamestmartin** [comment]
Actually actually I probably want to rewrite the whole thing to use local storage and iframes like Stack Overflow uses. It'd be more reliable and more secure.

**Comment by jamestmartin** [comment]
It looks like Stack Exchange's method *doesn't actually work* on either of the browsers I tested it on. 

Here's my new flow:

1. When the user logs into DAD, they receive a `SameSite=none; Secure` cookie with their cross-site auth code. This allows a cross-site enabled auth cookie without risking any XSRF attacks with the session cookie (because it's only applicable for this purpose).
2. JavaScript performs a cross-site GET with their x-auth code included as a cookie (possible because SameSite=none and CORS is enabled for that endpoint) to `/x_site_auth/auto_login_available` to check whether they are logged in to DAD (requires CORS). If not, it sets a short-lived cookie to avoid repeating this check until manual login is initiated.
3. JavaScript requests an auth code from the forums API.
4. JavaScript performs a cross-site POST with their auth code and the x-auth code included as a cookie to `/x_site_auth/verify`, which returns an HMAC-signed JWT which associates their forums auth code with their DAD account.
5. JavaScript POSTs the recieved JWT to the forums login endpoint, and is issued a forums session cookie. The user is now logged in.

Here's what the login button would do:

1. Check if automatic login is possible (ignoring the "don't spam auth" cookie). If so, log in. Otherwise, continue.
2. Delete the don't spam auth cookie.
3. Redirect the user to `/x_site_auth/login?return_to={:url}`. This page prompts the user to log in to DAD if they have not already (which is likely the case if they're seeing this page), and sets their x-auth code.
4. Redirect the user back to their original page, which will automatically log in because the "don't spam auth" cookie is not present.

Security issues *averted*:

* XSRF attacks on DAD: the main session cookie is still same-site, and CORS disables those requests anyway.
* XSRF attacks on the login endpoint specifically: CORS is only enabled for the forums origin. CORS is not enabled at all to generate the forums auth code.
* External sites spying on whether you're logged in to DAD: that endpoint requires CORS as well.
* Forums auth codes are short-lived and never present in the URL, preventing stealing.

Compared to the previous solution, this solution is:

* Easier on the user:
  * Auto-login can be performed without any redirects.
  * *Manual* login can be performed without any redirects if the user has logged into DAD since visiting the forums.
* More secure: No auth codes are ever present in URL query parameters.
* More reliable: Fewer weird 3rd-party cookies tricks.
* Probably easier to implement: the forums backend never has to directly query DAD in the login process (because of the JWT), and no changes need to be made to the session cookie.

File: app/controllers/x_site_auth_controller.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    # due to stuff like CORS settings, but they're still avoided to prevent client-side code bugs
+    # (i.e. so that *my* code won't set the wrong return URL).
+    unless @return_to.scheme == 'https'
+      flash[:danger] =
+        "HTTPS is required for cross-site authentication; #{@return_to} is not an HTTPS URL."
+      render_unauthorized
+      return
+    end
+
+    unless @return_to.host == ENV['X_AUTH_HOST']
```
Comment (the1banana) [inline app/controllers/x_site_auth_controller.rb:42]: What is X_AUTH_HOST? The website that DAD permits you to cross-authenticate login to (aka, only the DAD forums)?

Comment (jamestmartin) [reply app/controllers/x_site_auth_controller.rb:42]: Yes.

**Pull request merged**